java/ql/src/experimental/Security/CWE/CWE-094/TemplateInjection.qll

```diff
@@ -0,0 +1,241 @@
+        or
+        ma.getMethod() instanceof MethodJinjavaRender
+      )
+    )
+  }
+}
+
+/**
+ * An argument to ThymeLeaf template engine's `process` method call is as a sink for `Configuration`.
+ */
+private class ThymeLeafRenderSink extends Sink {
+  ThymeLeafRenderSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
+      ma.getMethod() instanceof MethodThymeleafProcess
+    )
+  }
+}
+
+private class ThymeLeafgetResourceAsStreamSink extends Sink {
+  ThymeLeafgetResourceAsStreamSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
+      ma.getMethod() instanceof MethodThymeleafGetResourceAsStream
+    )
+  }
+}
+
+/**
+ * An argument to MVEL template engine's `compileTemplate` method call is as a sink for `Configuration`.
+ */
+private class MVELCompileTemplateSink extends Sink {
+  MVELCompileTemplateSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
+      ma.getMethod() instanceof MethodMVELCompileTemplate
+    )
+  }
+}
+
+/**
+ * An argument to MVEL templating engine's `TemplateCompiler`
+ *   construtor call is a sink for `Configuration`.
+ */
+private class MVELConstructorSink extends Sink {
+  MVELConstructorSink() {
+    exists(ConstructorCall cc, Constructor c |
+      cc.getConstructor() = c and
+      c.getDeclaringType() instanceof MVELTemplateCompiler and
+      this.asExpr() = cc.getArgument(0)
+    )
+  }
+}
+
+/**
+ * Tainted data flowing into a Velocity Context through `put` method taints the context.
+ */
+private class VelocityContextFlow extends AdditionalFlowStep {
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
+    exists(MethodAccess m | m.getMethod() instanceof MethodVelocityContextPut |
+      m.getArgument(1) = prev.asExpr() and
+      (succ.asExpr() = m or succ.asExpr() = m.getQualifier())
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `mergeTemplate` method call is a sink for `Configuration`.
+ */
+private class VelocityMergeTempSink extends Sink {
+  VelocityMergeTempSink() {
+    exists(MethodAccess m |
+      // static boolean	mergeTemplate(String templateName, String encoding, Context context, Writer writer)
+      m.getCallee() instanceof MethodVelocityMergeTemplate and
+      m.getArgument(2) = this.getExpr()
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `evaluate` method call is a sink for `Configuration`.
+ */
+private class VelocityEvaluateSink extends Sink {
+  VelocityEvaluateSink() {
+    exists(MethodAccess m |
+      m.getCallee() instanceof MethodVelocityEvaluate and
+      m.getArgument([0, 3]) = this.getExpr()
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `parse` method call is a sink for `Configuration`.
+ */
+private class VelocityParseSink extends Sink {
+  VelocityParseSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
```

Comment:
Would it make sense to add the second argument (`Template`) as a sink too? In that case, we would need a taint step to taint a `Template` from the `Context` argument in its [`merge`](https://velocity.apache.org/engine/2.3/apidocs/org/apache/velocity/Template.html#merge-org.apache.velocity.context.Context-java.io.Writer-) method, though.

Reply:
It would result in a duplicate alert. The second arg can only be tainted by a call to one of the already existing sinks.

Reply:
It would result in a duplicate alert. The second argument can only be tainted by a call to one of the already existing sinks.